tests/scanner/audit/retention_rules_engine_test.py

```diff
@@ -0,0 +1,626 @@
+from tests.unittest_utils import ForsetiTestCase
+from google.cloud.forseti.scanner.scanners import retention_scanner
+
+do_not_test_old_cases = False
+
+
+def get_expect_violation_item(res_map, bucket_id, rule_name, rule_index):
+    RuleViolation = namedtuple(
+    'RuleViolation',
+    ['resource_name', 'resource_type', 'full_name', 'rule_name',
+     'rule_index', 'violation_type', 'violation_data', 'resource_data'])
+    lifecycle_str = json.dumps(res_map.get(bucket_id).get_lifecycle_rule())
+
+    return RuleViolation(
+        resource_name=bucket_id,
+        resource_type=res_map.get(bucket_id).type,
+        full_name=res_map.get(bucket_id).full_name,
+        rule_name=rule_name,
+        rule_index=rule_index,
+        violation_type=rre.VIOLATION_TYPE,
+        violation_data=lifecycle_str,
+        resource_data=res_map.get(bucket_id).data)
+
+def GetLefecycleDict(action, age, created_before, matches_storage_class,
+                     num_newer_versions, is_live):
+    result = {'action':{}, 'condition':{}}
+    result['action']['type'] = action
+    if age != None:
+        result['condition']['age'] = age
+    if created_before != None:
+        result['condition']['createdBefore'] = created_before
+    if matches_storage_class != None:
+        result['condition']['matchesStorageClass'] = matches_storage_class
+    if num_newer_versions != None:
+        result['condition']['numNewerVersions'] = num_newer_versions
+    if is_live != None:
+        result['condition']['isLive'] = is_live
+    return result
+
+_fake_bucket_list_22 = []
+
+def _mock_gcp_resource_iter_22(_, resource_type):
+    """Creates a list of GCP resource mocks retrieved by the scanner."""
+    """Used in test case 22"""
+    if resource_type != 'bucket':
+        raise ValueError(
+            'unexpected resource type: got %s, bucket',
+            resource_type,
+        )
+
+    return _fake_bucket_list_22
+
+def generate_res_for_22():
+    """Creates a list of GCP resource mocks retrieved by the scanner."""
+    """Used in test case 22"""
+
+    org = organization.Organization(
+        organization_id='433655555569',
+        full_name='organization/433655555569/')
+    proj_max = project.Project(
+        project_id='project-max',
+        project_number='711111111111',
+        full_name=org.full_name+'project/project-max/',
+        parent=org
+    )
+    proj_min = project.Project(
+        project_id='project-min',
+        project_number='722222222222',
+        full_name=org.full_name+'project/project-min/',
+        parent=org
+    )
+    proj_both = project.Project(
+        project_id='project-both',
+        project_number='733333333333',
+        full_name=org.full_name+'project/project-both/',
+        parent=org
+    )
+
+    res = []
+
+    # MAX
+    normal = 365
+    larger = 366
+
+    data_creater = frsd.FakeBucketDataCreater('bkt_max_1', proj_max)
+    data_creater.AddLefecycleDict("Delete", normal, None, None, None, None)
```

Comment:
For a method like this, give all the non-required arguments a default value of None, then use named arguments:
```
  AddLifecycleDict(action='Delete', age=499)
  AddLifecycleDict(action='Delete', age=500, num_newer_versions=5)
  ...
```